### PR TITLE
[InstCombine] Extend `foldICmpBinOp` to `add`-like `or`.

### PR DESCRIPTION
**Pull request by mgudim**
InstCombine canonicalizes `add` to `or` when possible, but this makes some optimizations applicable to `add` to be missed because they don't realize that the `or` is equivalent to `add`.

In this patch we generalize `foldICmpBinOp` to handle such cases.

### CODE REVIEW

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 1]
```diff
@@ -4564,6 +4564,17 @@ static Instruction *foldICmpXorXX(ICmpInst &I, const SimplifyQuery &Q,
   return nullptr;
 }
```
Comment (mgudim) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:4566]: I have seen checks for `or` which is actually an `add` in other places. Should we make this into a centralized utility function?

Comment (mgudim) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:4566]: Great, looks like this comes up in other places too:  https://github.com/llvm/llvm-project/pull/72583  
I'll wait for that PR to land, so I can apply it here.

**Comment by dtcxzyw** [comment]
Alive2: https://alive2.llvm.org/ce/z/B_7XNs

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp (hunk 1 above)
Comment (mgudim) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp]: Also see similar PR: https://github.com/llvm/llvm-project/pull/75899

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 2]
```diff
@@ -4624,31 +4624,37 @@ Instruction *InstCombinerImpl::foldICmpBinOp(ICmpInst &I,
...
-  if (BO1 && BO1->getOpcode() == Instruction::Add) {
-    C = BO1->getOperand(0);
-    D = BO1->getOperand(1);
+  if (BO1) {
+    match(BO1, m_AddLike(m_Value(C), m_Value(D)));
+    NoOp1WrapProblem = hasNoWrapProblem(*BO1, Pred, Op1HasNSW, Op1HasNUW);
   }
 
-  // icmp (A+B), A -> icmp B, 0 for equalities or if there is no overflow.
-  // icmp (A+B), B -> icmp A, 0 for equalities or if there is no overflow.
```
[the remark this answers is not in the document]
Comment (mgudim) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp]: My bad, thanks. Put this back in.

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 3]
```diff
@@ -4624,31 +4624,37 @@ Instruction *InstCombinerImpl::foldICmpBinOp(ICmpInst &I,
...
-
+  bool Op0HasNUW = false, Op1HasNUW = false;
+  bool Op0HasNSW = false, Op1HasNSW = false;
   // Analyze the case when either Op0 or Op1 is an add instruction.
   // Op0 = A + B (or A and B are null); Op1 = C + D (or C and D are null).
   Value *A = nullptr, *B = nullptr, *C = nullptr, *D = nullptr;
-  if (BO0 && BO0->getOpcode() == Instruction::Add) {
-    A = BO0->getOperand(0);
-    B = BO0->getOperand(1);
+  auto hasNoWrapProblem = [](const BinaryOperator &BO, CmpInst::Predicate Pred,
```
[the remark this answers is not in the document]
Comment (mgudim) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:4631]: done

File: llvm/test/Transforms/InstCombine/icmp.ll [hunk 4]
```diff
@@ -3862,10 +3862,9 @@ define <8 x i1> @bitreverse_vec_ne(<8 x i16> %x, <8 x i16> %y) {
 define i1 @knownbits1(i8 %a, i8 %b) {
 ; CHECK-LABEL: @knownbits1(
 ; CHECK-NEXT:    [[A1:%.*]] = and i8 [[A:%.*]], 1
-; CHECK-NEXT:    [[A2:%.*]] = or disjoint i8 [[A1]], 4
 ; CHECK-NEXT:    [[B1:%.*]] = and i8 [[B:%.*]], 2
-; CHECK-NEXT:    [[B2:%.*]] = or disjoint i8 [[B1]], 5
-; CHECK-NEXT:    [[C:%.*]] = icmp eq i8 [[A2]], [[B2]]
```
[the remark this answers is not in the document]
Comment (mgudim) [reply llvm/test/Transforms/InstCombine/icmp.ll:3868]: done.

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 5]
```diff
@@ -4624,27 +4624,35 @@ Instruction *InstCombinerImpl::foldICmpBinOp(ICmpInst &I,
...
-  if (BO1 && isa<OverflowingBinaryOperator>(BO1))
-    NoOp1WrapProblem =
-        ICmpInst::isEquality(Pred) ||
-        (CmpInst::isUnsigned(Pred) && BO1->hasNoUnsignedWrap()) ||
-        (CmpInst::isSigned(Pred) && BO1->hasNoSignedWrap());
-
+  bool Op0HasNUW = false, Op1HasNUW = false;
+  bool Op0HasNSW = false, Op1HasNSW = false;
   // Analyze the case when either Op0 or Op1 is an add instruction.
   // Op0 = A + B (or A and B are null); Op1 = C + D (or C and D are null).
```
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:4630]: This comment should be below the lambda (above the `Value *A = nullptr, *B = nullptr, *C = nullptr, *D = nullptr;` line).

**Review by nikic — APPROVED** [review]
LGTM

**Pull request merged**